pkg/kds/v2/store/sync.go

```diff
@@ -133,18 +134,25 @@ func (s *syncResourceStore) Sync(syncCtx context.Context, upstreamResponse clien
 	}
 	log.V(1).Info("before filtering", "downstream", downstream, "upstream", upstream)
 
+	filterResources := func(resources core_model.ResourceList, predicate func(core_model.Resource) bool) (core_model.ResourceList, error) {
+		return filter(resources, func(r core_model.Resource) bool {
+			_, exists := indexedInvalidResources[model.MetaToResourceKey(r.GetMeta())]
+			return !exists && predicate(r)
+		})
+	}
+
+	predicate := func(r core_model.Resource) bool { return true }
 	if opts.Predicate != nil {
-		if filtered, err := filter(downstream, opts.Predicate); err != nil {
-			return err
-		} else {
-			downstream = filtered
-		}
-		if filtered, err := filter(upstream, opts.Predicate); err != nil {
-			return err
-		} else {
-			upstream = filtered
-		}
+		predicate = opts.Predicate
+	}
```

Comment:
Another option could be to change filter to simply take a list of predicates. This might make this more readable